src/fundus/parser/utility.py

```diff
@@ -453,7 +447,95 @@ def parse_urls_from_image(node: lxml.html.HtmlElement) -> Optional[Dict[str, str
         return None
 
 
-_relative_source_selector = XPath("./ancestor::picture/source")
+class _DimensionCalculator:
+    def __init__(
+        self, width: Optional[float] = None, height: Optional[float] = None, ratio: Optional[float] = None
+    ) -> None:
+        self.width = width
+        self.height = height
+        self.ratio = ratio
+
+    def calculate(
+        self, width: Optional[float] = None, height: Optional[float] = None, dpr: Optional[float] = None
+    ) -> Optional[Dimension]:
+        if not (width or height):
+            width = self.width
+            height = self.height
+        if dimension := Dimension.from_ratio(width, height, self.ratio):
+            return dimension * (dpr or 1)
+        return None
+
+
+_min_width_pattern = re.compile(r"min-width:\s*(?P<minwidth>[0-9]*)(?P<descriptor>[pxem]+)")
+_width_x_height_pattern = re.compile(r"(?P<width>[0-9]+)x(?P<height>[0-9]+)")
+
+
+def get_versions_from_node(
+    source: lxml.html.HtmlElement, ratio: Optional[float], size_patter: Optional[Pattern[str]]
+) -> Set[ImageVersion]:
+    if not (urls := parse_urls(source)):
+        return set()
+
+    # get min width
+    min_width = None
+    if media := source.get("media"):
+        for value, descriptor in re.findall(_min_width_pattern, media):
+            if descriptor != "px":
+                logger.debug(f"Pixel calculation not implemented for {descriptor}")
+            else:
+                min_width = int(value)
+
+    # get width, height and init calculator
+    width = float(source.get("width", 0)) or None
+    height = float(source.get("height", 0)) or None
+    if width and height:
+        ratio = width / height
+    calculator = _DimensionCalculator(width, height, ratio)
+
+    versions = set()
+    for descriptor, url in urls.items():
+        kwargs: Dict[str, float] = {}
+        if descriptor is not None:
+            if match := re.search(r"(?P<multiplier>[0-9.]+)x", descriptor):
+                kwargs["dpr"] = float(match.group("multiplier"))
+            elif match := re.search(r"(?P<width>[0-9]+)(px|w)", descriptor):
```

Comment:
Other technically allowed units of measurement are `%` and `em`, also `w` doesn't seem to be a valid unit (https://www.w3schools.com/cssref/css_units.php)

Reply:
The URLs coming are coming from the `srcset` attribute. Allowed descriptors are `x`, `w`, I also found some `px`